include/flatbuffers/grpc.h

```diff
@@ -22,49 +22,198 @@
+    return slice_;
+  }
+
+  inline bool Verify() const {
+    if (slice_.refcount != nullptr) {
+      Verifier verifier(slice_.data.refcounted.bytes, slice_.data.refcounted.length);
+      return verifier.VerifyBuffer<T>(nullptr);
+    } else {
+      Verifier verifier(slice_.data.inlined.bytes, slice_.data.inlined.length);
+      return verifier.VerifyBuffer<T>(nullptr);
+    }
+  }
+
+  inline const T* GetRoot() const {
+    if (slice_.refcount != nullptr) {
+      return flatbuffers::GetRoot<T>(slice_.data.refcounted.bytes);
+    } else {
+      return flatbuffers::GetRoot<T>(slice_.data.inlined.bytes);
+    }
+  }
+
+ protected:
+  grpc_slice slice_;
+};
+
+// This isnt a true allocator, as it only allows up to two oustanding
+// allocation at a time. However, this is sufficient for our use case, and
+// simplifies the grpc_slice tracking.
+class SliceAllocator : public simple_allocator {
+ public:
+  inline SliceAllocator()
+      : slices_{grpc_empty_slice(), grpc_empty_slice()} {
+  }
+
+  SliceAllocator(const SliceAllocator& other) = delete;
+  SliceAllocator& operator=(const SliceAllocator& other) = delete;
+
+  virtual inline ~SliceAllocator() {
+    grpc_slice_unref(slices_[0]);
+    grpc_slice_unref(slices_[1]);
+  }
+
+  virtual inline uint8_t *allocate(size_t size) const override {
+    if (GRPC_SLICE_IS_EMPTY(slices_[0])) {
+      slices_[0] = grpc_slice_malloc(size);
+      return GRPC_SLICE_START_PTR(slices_[0]);
+    } else if (GRPC_SLICE_IS_EMPTY(slices_[1])) {
+      slices_[1] = grpc_slice_malloc(size);
+      return GRPC_SLICE_START_PTR(slices_[1]);
+    } else {
+      throw std::logic_error("SliceAllocator full");
+    }
+  }
+
+  virtual inline void deallocate(uint8_t *p) const override {
+    if (p == GRPC_SLICE_START_PTR(slices_[0])) {
+      grpc_slice_unref(slices_[0]);
+      slices_[0] = grpc_empty_slice();
+    } else if (p == GRPC_SLICE_START_PTR(slices_[1])) {
+      grpc_slice_unref(slices_[1]);
+      slices_[1] = grpc_empty_slice();
+    } else {
+      throw std::logic_error("invalid pointer to deallocate");
+    }
+  }
+
+  inline const grpc_slice& borrowslice(uint8_t *p) const {
+    if (p == GRPC_SLICE_START_PTR(slices_[0])) {
+      return slices_[0];
+    } else if (p == GRPC_SLICE_START_PTR(slices_[1])) {
+      return slices_[1];
+    } else {
+      throw std::logic_error("invalid pointer to deallocate");
+    }
+  }
+
+ protected:
+  mutable grpc_slice slices_[2];
+};
+
+class MessageBuilder : public SliceAllocator, public FlatBufferBuilder {
+ public:
+  inline MessageBuilder(uoffset_t initial_size = 1024)
+      : SliceAllocator(),
+        FlatBufferBuilder(initial_size, this) {
+  }
+
+  MessageBuilder(const MessageBuilder& other) = delete;
+  MessageBuilder& operator=(const MessageBuilder& other) = delete;
+
+  inline ~MessageBuilder() {
+  }
+
+  template <class T>
+  inline Message<T> GetMessage() const {
+    assert(buf_.data() > buf_.buf());
```

Comment:
Yeah, I can document this inline with the code. Essentially this extracts a subslice of a `grpc_slice` that shares the refcount, so that we don't need to make a copy to reference the message we are sending out.